httprule/httprule.go

```diff
@@ -0,0 +1,77 @@
+// Package httprule provides utilities to map google.api.http annotation
+// to net/http Request and Response types. These utilities allow to
+// generate HTTP Clients for a given proto service. The methods of this
+// service have their HTTP mappings specified `via google.api.http`
+// method options, e.g.:
+//
+// service HelloService {
+//     rpc Hello (HelloRequest) returns (HelloResponse) {
+//         option (google.api.http) = { post:"/api/hello" body:"*" };
+//     };
+// };
+
+// HttpRule proto:   https://github.com/googleapis/googleapis/blob/master/google/api/http.proto
+// HttpRule codegen: https://pkg.go.dev/google.golang.org/genproto/googleapis/api/annotations
+package httprule
+
+import (
+	"fmt"
+	"io"
+
+	pb "google.golang.org/genproto/googleapis/api/annotations"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+)
+
+var (
+	ErrInvalidHttpRule = fmt.Errorf("invalid HttpRule")
+)
+
+// ParseProtoResponse parses a proto message from a HTTPRule and an
+// io.Reader, typically an http.Response.Body. rule.ResponseBody may
+// only reference a top-level field of the response target proto
+// message.
+func ParseProtoResponse(rule *pb.HttpRule, r io.Reader, respTarget proto.Message) error {
+	if err := ValidateHTTPRule(rule); err != nil {
+		return err
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return fmt.Errorf("cannot parse proto response from reader: %w", err)
+	}
+	if rule.ResponseBody != "" {
+		respTarget, err = newField(rule.ResponseBody, respTarget)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := protojson.Unmarshal(b, respTarget); err != nil {
+		return fmt.Errorf("cannot protojson unmarshal: %w", err)
+	}
+	return nil
+}
+
+func newField(fieldName string, msg proto.Message) (proto.Message, error) {
+	m := msg.ProtoReflect()
+	fds := m.Descriptor().Fields()
+	fd := fds.ByTextName(fieldName)
+	if fd == nil || fd.Message() == nil {
+		return nil, fmt.Errorf("cannot find message-typed, top-level field '%s': %w", fieldName, ErrInvalidHttpRule)
+	}
+
+	sub := m.Get(fd).Message().New().Interface() // WTF?!??? couldn't find another way
```

Comment:
I need to study this more in the context of the protoreflect apis - I don't really understand what is going on with this and the next couple of lines.